accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+Wasm; we just need a debug adapter component to bridge the gap between the
+high-level source language and low-level Wasm semantics.
+
+We will implement two adapters:
+
+1. **The Null Debug Adapter:** A debug adapter that forwards all queries
+   directly to the underlying Wasm being executed. This provides Wasm-level
+   debugging to the developer.
+
+2. **The DWARF Debug Adapter:** A debug adapter component that translates
+   queries from the debugger through the source program’s DWARF debug info
+   custom sections. This enables the debugging of programs compiled to Wasm,
+   such as Rust, C, C++, etc… 
+
+Interpreters that are compiled to Wasm can, in the future, provide their own
+debug adapter components that translate debugging queries through their runtime
+data structures to inspect debuggee programs being interpreted within
+themselves.
+
+Even further into the future, we can collaborate with the WebAssembly Community
+Group to standardize debug adapter components (or something evolved from them)
+in the ``WebAssembly/tool-conventions`` and `WebAssembly/component-model`
+repositories.
+
+While the exact WIT interface of and world for debug adapter components is out
+of scope for this RFC, and will require experimentation and prototyping during
+implementation, we expect that it will closely mirror the kinds of queries
+supported by the Debug Adapter Protocol.
+
+## Live Debugging
+
+Live debugging support comprises several features, each of which comes with
+design and implementation considerations. These features are: setting
+breakpoints, single stepping, watch points, step into/out of, and integration
+with our existing baseline compiler Winch.
+
+### Breakpoints and Single-stepping
+
+To implement breakpoints and single-stepping, we need the ability to pause
+program execution and wait for user input prior to executing a specific line of
+code. The way this is handled in native debuggers is as follows:
+
+* When a user sets a breakpoint, the instruction corresponding to that line of
+  code is replaced with an `int3` instruction.
+
+* When executed, the `int3` instruction raises a `SIGTRAP`, halting execution
+  and yielding control back to the debugger.
+
+This approach leads to a problem for us: we only have one compiled copy of a
+module’s code, even though multiple instances may be running the code. So if we
+just replace an instruction in the code with an interrupt, it is possible that a
+different instance will hit that interrupt. At that point, in order for that
+instance to continue executing we need to swap in the actual instruction. This
+leads to a race condition between the instances where we aren’t guaranteed that
+our breakpoint will actually interrupt when we need it to! We can get around
+this in the following way:
+
+1. When code is compiled for debugging, for every instruction emitted, emit a
+call to a utility function that can check if execution should be paused:
```

Comment:
Another thing that we could possibly do, if we're using DWARF, is to only insert these calls for every source line we encounter.  This would still allow for line-by-line debugging while improving performance of the debugged code.

Reply:
Instruction level debugging can be quite useful when debugging UB. Maybe it could be an option to enable or disable?

Reply:
The other thing that this would preclude is switching between source-level and instruction-level debugging on-the-fly during the same debugging session, which is rarely needed but crucial when you do need it.

So I agree that this could be an option (eventually).

But I think we should really focus on making debugging simple and correct. Nothing worse than a buggy debugger that lies to you! So let's focus first on correctness and worry about the performance of debuggable code second. Not to say we shouldn't consider it at all, but as long as we don't box ourselves in architecturally, it should be able to safely set such questions aside for the most part.